qiskit/pulse/library/symbolic_pulses.py

```diff
@@ -0,0 +1,754 @@
+    After providing the envelope definition, you can generate the waveform data.
+    Note that it would be convenient to define a factory function that automatically
+    accomplishes this procedure.
+
+    .. code-block:: python
+
+        def Sawtooth(duration, amp, freq, name):
+            instance = SymbolicPulse(
+                pulse_type="Sawtooth",
+                parameters={"duration": duration, "amp": amp, "freq": freq},
+                name=name,
+            )
+
+            t, amp, freq = sympy.symbols("t, amp, freq")
+            instance.envelope = amp * 2 * (freq * t - sympy.floor(1 / 2 + freq * t))
+
+            return instance
+
+    You can also provide a :class:`Parameter` object in the ``parameters`` dictionary
+    when you instantiate the symbolic pulse instance. Waveform cannot be
+    generated until you assign all unbounded parameters.
+    Note that parameters will be assigned through the schedule playing the pulse.
+
+
+    .. _symbolic_pulse_serialize:
+
+    .. rubric:: Serialization
+
+    The :class:`~SymbolicPulse` subclass is QPY serialized with symbolic expressions.
+    A user can therefore create a custom pulse subclass with a novel envelope and constraints,
+    and then one can instantiate the class with certain parameters to run on a backend.
+    This pulse instance can be saved in the QPY binary, which can be loaded afterwards
+    even within the environment not having original class definition loaded.
+    This mechanism allows us to easily share a pulse program including custom pulse instructions
+    with collaborators or to directly submit the program to the quantum computer backend
+    in the parametric form (i.e. not sample data). This greatly reduces amount of data transferred.
+    The waveform sample data to be loaded in the control electronics
+    will be reconstructed with parameters and deserialized expression objects by the backend.
+
+    .. note::
+
+        Currently QPY serialization of :class:`SymbolicPulse` is not available.
+        This feature will be implemented shortly.
+        Note that data transmission in the parametric form requires your quantum computer backend
+        to support QPY framework with :class:`SymbolicPulse` available. Otherwise, the pulse data
+        might be converted into :class:`Waveform`.
+    """
+
+    # Lambdify caches keyed on sympy expressions. Returns the corresponding callable.
+    _callable_envelope = LamdifiedExpression(
+        common_params=["t", "duration", "amp"],
+        attribute="_envelope_expr",
+    )
+    _callable_consts = LamdifiedExpression(
+        common_params=["duration", "amp"],
+        attribute="_consts_expr",
+    )
+
+    def __init__(
+        self,
+        pulse_type: str,
+        parameters: Dict[str, Union[ParameterExpression, complex]],
+        name: Optional[str] = None,
+        limit_amplitude: Optional[bool] = None,
+    ):
+        """Create a parametric pulse and validate the input parameters.
```

Comment:
This is because I assumed the case that the class is instantiated without expressions, i.e. validation cannot be performed without expression. So user should call validation by themself, and they can also avoid validation for performance.

Reply:
Good catch! thanks